# Examen Adelantado Econometría 2020-2021
## 12 de Enero de 2021


### Instrucciones

+ La duración del examen de 1.5 horas.
+ Cada pregunta se responderá en el recuadro bajo la pregunta, bien en modo sólo texto (Preguntas 1 a 5) o bien en modo códigos de Python + texto respuesta en base a los resultados obtenidos (Preguntas 6 a 10).
+ Los datos necesarios está disponibles bien en la librería wooldridge o en el fichero Excel "Pregunta7.xls" que se encuentra en github (las órdenes orientativas para leer los ficheros se encuentran descritas bajo cada uno de esos ejercicios).
+ No se evaluarán respuestas en las que únicamente haya códigos. Los resultados deberán servir para responder a las preguntas.
+ Las puntuaciones de cada ejercicio se encuentran en el propio enunciado.
+ El fichero final del examen se enviará a vía PRADO en la Tarea habilitada para el examen.


### APELLIDOS,  NOMBRE: JOSÉ LUIS GUIDÚ NAVAS
### DNI: 79146348J

### Pregunta 1: (1 punto)

Justificar que $\widehat{\boldsymbol{\beta}} \sim N \left( \boldsymbol{\beta}, \ \sigma^{2} \cdot \left( \mathbf{X}^{t} \mathbf{X} \right)^{-1} \right)$. ¿Qué supone este resultado cuando realizamos inferencia sobre los coeficientes de un modelo lineal múltiple?

- Respuesta:


### Pregunta 2: (1 punto)

Considera el modelo $Y_i = \beta_0 + \beta_1 X_i + u_i$ ($i=1,2,\ldots,20$). Si realizamos la regresión de Mínimos Cuadrados Ordinarios de los residuos obtenidos en el modelo al cuadrado $u_i^2$ sobre las variables $X$ y $X^2$ (y la constante), obtenemos un $R^2$ de 0.35.  ¿Qué podemos decir sobre la homocedasticidad del modelo planteado?

- Respuesta:

Como el modelo no se ajusta bien a los datos dados debido a que el R^2 es un valor bajo.

### Pregunta 3: (0.75 puntos)

Enumerar y describir brevemente la hipótesis básicas necesarias para realizar el análisis de un
modelo de regresión lineal múltiple. Justificar las razones por las que se exigen cada una de
estas hipótesis.

- Respuesta:

Hipótesis básicas sobre el modelo:
✠ El supuesto de linealidad. Nos deja utilizar procedimientos algebraicos sobre matrices y vectores simplificando su operatoria.
✠ El supuesto rango completo por columnas: En el caso de que no se verifique habría problema de multicolinealidad.
el supuesto mencionado, implicaría que alguna de las variables explicativas sería combinación lineal de otras de las variables, produciéndose un problema de multicolinealidad.
✠ El supuesto de exogeneidad. Si el valor esperado de la perturbación es cero.
✠ El supuesto de causalidad. El mecanismo de generación de las observaciones.
✠ Supuestos sobre el término perturbación. Nos deja saber si estamos ante un problema de heterocedasticidad o de
multicolinealidad.
✠ El supuesto de normalidad del término de perturbación. Realiza una inferencia sobre los parámetros que intervienen en el modelo.


### Pregunta 4: (0.5 puntos)

En el caso de tener que decidir qu\'e modelo es el mejor en el siguiente caso:
        \begin{eqnarray*}
            \mbox{Modelo 1: } & & \mathbf{Y} = \beta_{1} + \beta_{2} \cdot \mathbf{X}_{2} + \beta_{3} \cdot \mathbf{X}_{3} + \mathbf{u}. \\
            \mbox{Modelo 2: } & & \mathbf{Y} = \alpha_{1} + \alpha_{2} \cdot \mathbf{X}_{2} + \alpha_{3} \cdot \mathbf{X}_{3} + \alpha_{4} \cdot \mathbf{X}_{4} + \mathbf{v}.
        \end{eqnarray*}
        ¿Qué herramientas usaría y con qué criterios? Justifique su respuesta.

- Respuesta:

Tendríamos que calcular el AIC y el BIC de ambos modelos. Una vez calculados nos fijaremos en que modelo tiene el AIC y el BIC mas bajo. El modelo cuyo AIC y BIC fuese inferior, sería el mejor modelo. Se usa esta herramienta porque otras herramientas de ajuste como el R^2 no tienen en cuenta el número de variables por ejemplo.


### Pregunta 5: (0.75 puntos)

Consideremos un modelo económico en el que se desea explicar el precio de la vivienda en función de la zona donde se encuentra la vivienda, sus metros cuadrados, número de habitaciones, número de baños y si tiene o no cochera. Puesto que se sospecha que pudieran existir relaciones lineales entre los metros cuadrados de la vivienda y el número de baños y habitaciones de la misma, se calcula el factor de inflación de la varianza, obteniéndose que el mayor de ellos es igual a 4.78. ¿Se puede decir que el grado de multicolinealidad aproximada existente es preocupante? 

- Respuesta:
No, la multicolinealidad no es preocupante ya que el FIV es menor a 10.


### Pregunta 6: (1.5 puntos)

Se desea analizar el número de becarios en la enseñanza universitaria, $\mathbf{B}$ (medidos en miles), a partir del número de alumnos matriculados en los estudios de primer y segundo ciclo, $\mathbf{A}$ (medidos en miles), y el importe destinado a becas, $\mathbf{I}$ (medido en millones de euros). A partir de las universidades de las 17 comunidades autónomas más la universidad a distancia, se obtiene:
        $$(\mathbf{X}^{t} \mathbf{X})^{-1} =  \begin{pmatrix} 0.114 & -0.0005 & -0.0002 \\ -0.0005 & 0.00002 & -0.00002 \\ -0.0002 & -0.00002 & 0.00003 \end{pmatrix}, \mathbf{X}^{t} \mathbf{y} = 
            \begin{pmatrix}
                443.818 \\
         66792.89 \\
                53125.748
            \end{pmatrix}, \mathbf{y}^{t} \mathbf{y} = 21953.5,$$
            
Se pide responder de forma razonada las siguientes cuestiones:
1. Especificar los modelos económico y econométrico.
1. Obtener la estimación de los coeficientes del modelo.
1. ¿Es el coeficiente del número de alumnos matriculados en los estudios de primer y segundo ciclo significativamente distinto de cero al 95% de confianza? Indique las consecuencias del análisis realizado.
1. Es el coeficiente del importe destinado a becas significativamente distinto de cero al 5% de significación? Indique las consecuencias del análisis realizado.
1. Es el modelo globalmente significativo? 

In [ ]:
#Cálculos:


- Respuesta:


### Pregunta 7: (1.5 puntos)

Se considera el modelo econométrico $Y_i = \beta_0 X_i^{\beta_1} X_2^{\beta_2} {\rm exp}({u_i})$ $i=1, \ldots, n$, donde $Y$ es la producción total de un bien, $X_1$ es el stock de capital, $X_2$ es el número de empleados para producir el bien y $u_i$ es la perturbación aleatoria. 
1. Transforma el modelo anterior en un modelo lineal múltiple.
1. ¿Cómo se interpretarían los coeficientes del modelo? 
1. Que sentido tendría que $\beta_1+\beta_2=1$.
1. Estima el modelo para los datos de "Pregunta7.xls", interpreta los coeficientes obtenidos y comprueba si $\beta_1+\beta_2=1$.

In [11]:
#Cálculos:

import pandas as pd

datos7 = pd.read_excel ("Pregunta7.xls")
#print(datos7)

#Tomamos logaritmos
datos7["lY"] = np.log(datos7["Y"])
datos7["lX1"] = np.log(datos7["X1"])
datos7["lX2"] = np.log(datos7["X2"])

y=datos7["lY"]
X=datos7[["lX1","lX2"]]

mco7=sm.OLS(y, sm.add_constant(X)).fit()
mco7.summary()

#Para realizar tests de hipotesis R*B=r
mco7.f_test(([0,1,1],1))
#Al no poder rechazar la hip. nula porque el p-valor es más grande que 0.05. Damos por valido que 𝛽1+𝛽2=1.

<class 'statsmodels.stats.contrast.ContrastResults'>
<F test: F=array([[0.19558359]]), p=0.6628307009410124, df_denom=21, df_num=1>

- Respuesta:
Lo transformamos en lineal múltiple tomando logaritmos.
Yi = B0 * Xi^B1 * X2^B2 * exp(ui)
log(Yi) = log(B0 * Xi^B1 * X2^B2 * exp(ui))
log(Yi) = log(B0) + log(Xi^B1) + log(X2^B2) + log(exp(ui))
log(Yi) = log(B0) + B1*log(Xi) + B2*log(X2) + ui

Al tener log en ambos miembros: Un aumento de un 1% de la X afecta un beta% a la Y.

Si la suma de los beta es 1, un incremento conjunto de los X de 1% aumenta un 1% la Y

Sabiendo que 𝑌  es la producción total de un bien, 𝑋1 es el stock de capital, 𝑋2 es el número de empleados para producir el bien, observando el estudio del modelo podemos ver como es globalmente significativo con un p-valor inferior a 0.05. También son significativas individualmente lX1 y lX2 por el p-valor inferior a 0.05. Diríamos que un 1% de variación en X1 aumenta 0.2331% la Y, lo mismo con X2 pero con un incremento del 0.8073%.
Después de realizar el test de hipótesis, al no poder rechazar la hip. nula porque el p-valor es más grande que 0.05. Damos por valido que 𝛽1+𝛽2=1.

### Pregunta 8: (1 punto)

En la base de datos $\texttt{phillips}$ se encuentra información para analizar el efecto del desempleo (unem) en el IPC (inf) desde 1948 a 2003. Estima el modelo lineal simple y estudia la existencia de autocorrelación en éste. En caso afirmativo, aplica alguna estrategia para corregirla y comprueba si ha concluido con éxito.

In [13]:
#Cálculos:

from wooldridge import *
datos8=dataWoo("phillips")

#Cálculos:

from wooldridge import *
import statsmodels.api as sm
datos8=dataWoo("phillips")

Y=datos8["inf"] #la variable endogena será la columna sinf
X=sm.add_constant(datos8["unem"]) #añadimos constante a las variables exogenas

mco8=sm.OLS(Y, X).fit()  #ajustamos el modelo 
mco8.summary()  #resumen del modelo ajustado

#realizamos el contraste de durbin watson
from statsmodels.stats.stattools import durbin_watson 
dw=durbin_watson(mco8.resid)
dw    #dw es igual a 0.80 por lo que hay presencia de correlacion positiva

#corregimos el modelo con la modificación Prais-Winsten
rho= 1 - dw/2

mco8_autocorr=sm.GLSAR(Y, sm.add_constant(X), rho=rho).iterative_fit(maxiter=1000)
mco8_autocorr.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                           GLSAR Regression Results                           
==============================================================================
Dep. Variable:                    inf   R-squared:                       0.093
Model:                          GLSAR   Adj. R-squared:                  0.076
Method:                 Least Squares   F-statistic:                     5.436
Date:                Tue, 12 Jan 2021   Prob (F-statistic):             0.0236
Time:                        10:25:14   Log-Likelihood:                -118.95
No. Observations:                  55   AIC:                             241.9
Df Residuals:                      53   BIC:                             245.9
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          7.3837      2.224      3.321      0.002       2.924      11.844
unem          -0.6850      0.294     -2.331      0.024      -1.274      -0.096
==============================================================================
Omnibus:                        9.225   Durbin-Watson:                   1.626
Prob(Omnibus):                  0.010   Jarque-Bera (JB):               19.725
Skew:                          -0.230   Prob(JB):                     5.21e-05
Kurtosis:                       5.897   Cond. No.                         12.0
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

- Respuesta:


- Respuesta: 
El modelo lineal simpre es IPC = 1.0536 + 0.05024*UNEM. Podemos afirmar la presencia de autocorrelacion positiva
ya que el contraste de Durbin-Watson nos da 0,801 que es cercano a 0. Aplicamos la modificacion de Prais-Winsten para corregirlo.
El nuevo modelo tiene un DW de 1.626 que esta mucho más cercano a 2 habiendo eliminado parte de la autocorrelacion.


### Pregunta 9: (1 punto)

Comprueba la existencia de heteroscedasticidad en el modelo $\log(price) = \beta_0+ \beta_1 \log(lotsize)+\log(sqrft)+bdrms$ usando los datos de $\texttt{hprice1}$ utilizando el test de Breush-Pagan. ¿Qué consecuencias tendría en el análisis del modelo su existencia?

In [4]:
#Cálculos:

from wooldridge import *
datos9=dataWoo("hprice1")
#datos9=dataWoo("hprice1", description=True)

from statsmodels.stats.diagnostic import het_breuschpagan
from statsmodels.stats.diagnostic import het_white
from statsmodels.formula.api import ols

import numpy as np
import pandas as pd #librería para manejo de datos
import statsmodels.api as sm
import statsmodels.formula.api as smf
import matplotlib.pylab as plt


y = datos9['lprice'] #Nota final es la variable endógena (la que quiero explicar)
X = datos9[['llotsize', 'lsqrft', 'bdrms']] #Variables exógenas


results = sm.OLS(y,sm.add_constant(X)).fit()

#Test de Breusch-Pagan
bp_test = het_breuschpagan(results.resid, results.model.exog)

labels = ['LM Statistic', 'LM-Test p-value', 'F-Statistic', 'F-Test p-value']
print(dict(zip(labels, bp_test)))

{'LM Statistic': 4.223248117304362, 'LM-Test p-value': 0.2383445905811877, 'F-Statistic': 1.411500740087146, 'F-Test p-value': 0.24514541735245635}


- Respuesta:
Procederemos a utilizar el test de **Breusch-Pagan**. Este test estudia si la variaza de los residuos depende de las variables independientes.
$$e^2 = \delta_1 x_1+ ... + \delta_p X_p+u$$
por lo que contrasta la hipótesis nula:
$$H_0:\delta_1=\delta_2=...=0$$
Si se rechaza la hipótesis nula existe heterocedasticidad.

Si el p-valor es superior a nuestro alpha=0.05 (lo es) entonces no podemos rechazar nuestra hipotesis nula y afirmamos la presencia de homocedasticidad.

### Pregunta 10: (1 punto)

Se pretende analizar el efecto de los ingresos per cápita (pcinc), de la población (popul) y del número de médicos por cada 100 mil habitantes (physic) en la mortalidad infantil (infmrt), medida como el número de muertes infantiles por cada mil nacimientos vivos.
1. Usando la base de datos $\texttt{infmrt}$ para 1990, estima el modelo econométrico lineal usando los logaritmos de las variables exógenas e interpreta los resultados obtenidos.
2. Analiza si existen diferencias significativas entre los resultados obtenidos para 1990 y para los de 1987.

In [16]:
#Cálculos:

from wooldridge import *
datos10=dataWoo("infmrt")

#Cálculos:
import numpy as np
from wooldridge import *
datos10=dataWoo("infmrt")


datos10["lpcinc"]=np.log(datos10["pcinc"])
datos10["lpopul"]=np.log(datos10["popul"])
datos10["lphysic"]=np.log(datos10["physic"])




Y=datos10['infmort'] #la variable endogena será la columna infmort
X=sm.add_constant(datos10[['lpcinc', 'lpopul','lphysic']]) #añadimos constante a las variables exogenas

mco10=sm.OLS(Y, X).fit()  #ajustamos el modelo 
print(mco10.summary())  #resumen del modelo ajustado
SCR=mco10.ssr  #SCR del total de los datos

#realizamos el test de Chow

datos1987=datos10[datos10["year"]==1987]  #modelo para 1987
Y1987=datos1987["infmort"]
X1987=datos1987[["lpcinc", "lpopul", "lphysic"]]
mco1987=sm.OLS(Y1987, sm.add_constant(X1987)).fit()
SCR1987=mco1987.ssr


datos1990=datos10[datos10["year"]==1990]  #modelo para 1990
Y1990=datos1990["infmort"]
X1990=datos1990[["lpcinc", "lpopul", "lphysic"]]
mco1990=sm.OLS(Y1990, sm.add_constant(X1990)).fit()
SCR1990=mco1990.ssr


n=len(datos10) 
k=3 #numero variables

Fexp=(n-2*k-2)/(k+1) * (SCR - (SCR1987+SCR1990))/(SCR1987+SCR1990)
Fexp

from scipy import stats
Fteo= stats.f.ppf(1-0.05, k+1, n-2*k-1)

print(Fexp, Fteo) #se rechaza que 1990 y 1987 son iguales si Fexp>Fteo

                            OLS Regression Results                            
Dep. Variable:                infmort   R-squared:                       0.182
Model:                            OLS   Adj. R-squared:                  0.157
Method:                 Least Squares   F-statistic:                     7.260
Date:                Tue, 12 Jan 2021   Prob (F-statistic):           0.000190
Time:                        10:28:36   Log-Likelihood:                -207.70
No. Observations:                 102   AIC:                             423.4
Df Residuals:                      98   BIC:                             433.9
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         36.2261     10.135      3.574      0.0

- Respuesta:


El modelo es globalmente significativo con un p-valor inferior a 0.05, las variables son significativas salvo lpopul que tiene un pvalor de 0.776. 

Un incremento del 1% en la variable pcinc hace que la Y aumente en -4.8841 unidades, si aumenta pcinc, la Y (mortalidad infantil) disminuye. 

Aunque no siendo logpopul significativa, un aumento del 1% en la variable popul hará que la Y aumente en -0.0536 unidades, si aumenta, la mortalidad infantil disminuirá. 

Un aumento del 1% en la variable pcinc hará que la Y aumente en 4.0278 unidades, si aumenta, la Y (mortalidad infantil) aumenta 
tambien, cabría destacar la paradoja que aparantemente es que cuantos más médicos más mortalidad infantil.

2)Realizamos el test de Chow:
La hipotesis nula es que la regresion lineal de la mortalidad de ambos años es la misma, obteniendo un Fexp=0,0332 y Fteo=2,467. Como Fexp<Fteorica no rechazamos la hipotesis nula. Por lo que podemos afirmar con una seguridad del 95% que no existen diferencias entre ambos años.